commands/build_bundles.py

```diff
@@ -0,0 +1,159 @@
+
+@retry_timeout
+def fetch_attachment(url):
+    print("Fetch %r" % url)
+    resp = requests.get(url)
+    return resp.content
+
+
+def write_zip(output_path: str, content: list[tuple[str, bytes]]):
+    """
+    Write a Zip at the specified `output_path` location with the specified `content`.
+    The content is specified as a list of file names and their binary content.
+    """
+    parent_folder = os.path.dirname(output_path)
+    os.makedirs(parent_folder, exist_ok=True)
+
+    zip_buffer = io.BytesIO()
+    with zipfile.ZipFile(zip_buffer, "w", zipfile.ZIP_DEFLATED) as zip_file:
+        for filename, filecontent in content:
+            zip_file.writestr(filename, filecontent)
+    with open(output_path, "wb") as f:
+        f.write(zip_buffer.getvalue())
+    print("Wrote %r" % output_path)
+
+
+def sync_cloud_storage(folder, storage_bucket):
+    """
+    Synchronizes a local folder (eg. `bundles/`) with a remote one in the specified
+    `storage_bucket` name.
+    """
+    # Ensure you have set the GOOGLE_APPLICATION_CREDENTIALS environment variable
+    # to the path of your Google Cloud service account key file before running this script.
+    client = storage.Client()
+    bucket = client.bucket(storage_bucket)
+    local_files = set()
+    for root, _, files in os.walk(folder):
+        for file in files:
+            local_file_path = os.path.join(root, file)
+            remote_file_path = os.path.join(folder, file)
+
+            blob = bucket.blob(remote_file_path)
+            blob.upload_from_filename(local_file_path)
+            print(f"Uploaded {local_file_path} to gs://{storage_bucket}/{remote_file_path}")
+            local_files.add(remote_file_path)
+
+    blobs = bucket.list_blobs(prefix=folder)
+    for blob in blobs:
+        if blob.name not in local_files:
+            blob.delete()
+            print(f"Deleted gs://{storage_bucket}/{blob.name}")
+
+
+def build_bundles(event, context):
+    """
+    Main command entry point that:
+    - fetches all collections changesets
+    - builds a `bundles/changesets.zip`
+    - fetches attachments of all collections with bundle flag
+    - builds `bundles/{bid}--{cid}.zip` for each of them
+    - synchronizes the `bundles/` folder with a remote Cloud storage bucket
+    """
+    rs_server = event.get("server") or SERVER
+
+    client = KintoClient(server_url=rs_server)
+
+    all_changesets = fetch_all_changesets(client)
+    write_zip(
+        f"{DESTINATION_FOLDER}/changesets.zip",
+        [
+            ("{bucket}--{metadata[id]}.json".format(**changeset), json.dumps(changeset))
+            for changeset in all_changesets
+        ],
+    )
+
+    base_url = client.server_info()["capabilities"]["attachments"]["base_url"]
+
+    for changeset in all_changesets:
+        if not BUILD_ALL and not changeset["metadata"].get("attachment", {}).get("bundle", False):
+            # Bundling not enabled.
+            continue
+
+        # Skip bundle if no attachments found.
```

Comment:
We should skip if the bundle already exists and is newer than the most recent change.

Reply:
Done !